src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -50,120 +52,119 @@ public Result charTyped(char charTyped,
                             @NotNull Project project,
                             @NotNull Editor editor,
                             @NotNull PsiFile file) {
-        LSPFormattingFeature formattingFeature = getFormattingFeature(file);
-        if (formattingFeature != null) {
-            boolean rangeFormattingSupported = formattingFeature.isRangeFormattingSupported(file);
-
-            // Close braces
-            if (formattingFeature.isFormatOnCloseBrace(file) &&
-                // Make sure the formatter supports formatting of the configured scope
-                ((formattingFeature.getFormatOnCloseBraceScope(file) == FormattingScope.FILE) || rangeFormattingSupported)) {
-                Map.Entry<Character, Character> bracePair = ContainerUtil.find(
-                        LSPIJEditorUtils.getBracePairs(file).entrySet(),
-                        entry -> entry.getValue() == charTyped
-                );
-                if (bracePair != null) {
-                    String formatOnCloseBraceCharacters = formattingFeature.getFormatOnCloseBraceCharacters(file);
-                    Character openBraceChar = bracePair.getKey();
-                    Character closeBraceChar = bracePair.getValue();
-                    if (StringUtil.isEmpty(formatOnCloseBraceCharacters) ||
-                        (formatOnCloseBraceCharacters.indexOf(closeBraceChar) > -1)) {
-                        return handleCloseBraceTyped(
-                                project,
-                                editor,
-                                file,
-                                formattingFeature,
-                                openBraceChar,
-                                closeBraceChar
-                        );
+        // Gather all of the relevant client configuration
+        BooleanRef rangeFormattingSupportedRef = BooleanRef.create();
+        BooleanRef formatOnCloseBraceRef = BooleanRef.create();
+        ComparableRef<FormattingScope> formatOnCloseBraceScopeRef = ComparableRef.create(FormattingScope.CODE_BLOCK);
+        StringRef formatOnCloseBraceCharactersRef = StringRef.create();
+        BooleanRef formatOnStatementTerminatorRef = BooleanRef.create();
+        ComparableRef<FormattingScope> formatOnStatementTerminatorScopeRef = ComparableRef.create(FormattingScope.STATEMENT);
+        StringRef formatOnStatementTerminatorCharactersRef = StringRef.create();
+        BooleanRef formatOnCompletionTriggerRef = BooleanRef.create();
+        StringRef formatOnCompletionTriggerCharactersRef = StringRef.create();
+        LanguageServiceAccessor.getInstance(project).processLanguageServers(
+                file,
+                ls -> {
+                    if (ls.getServerStatus() == ServerStatus.started) {
+                        LSPFormattingFeature formattingFeature = ls.getClientFeatures().getFormattingFeature();
```

Comment:
Collect all of that info across all matching, started language servers.

Reply:
You need to check that ls enable/support formatting features (and not on type formatting). See old hasLanguageServerSupportingOnlyFormatting

Reply:
Done.

Reply:
IMHO we should use hasAny and processLanguageServers both because you evaluate mappings twice. I wonder if your processLanguageServers should have a new parameter with filter.

Reply:
100% agreed. That was a sloppy oversight on my part. Running tests on a change to consolidate that check into the existing loop.

Reply:
Done.